diff -Naur a/build-aux/config.guess b/build-aux/config.guess
--- a/build-aux/config.guess	2013-02-22 17:56:24.000000000 +0400
+++ b/build-aux/config.guess	2024-02-03 13:57:20.000000000 +0100
@@ -950,6 +950,9 @@ EOF
     amd64:CYGWIN*:*:* | x86_64:CYGWIN*:*:*)
 	GUESS=x86_64-pc-cygwin
 	;;
+    amd64:MSYS*:*:* | x86_64:MSYS*:*:*)
+	GUESS=x86_64-pc-msys
+	;;
     prep*:SunOS:5.*:*)
 	SUN_REL=`echo "$UNAME_RELEASE" | sed -e 's/[^.]*//'`
 	GUESS=powerpcle-unknown-solaris2$SUN_REL
