diff --git a/Configurations/10-main.conf b/Configurations/10-main.conf
index b1719be..f68dc4e 100644
--- a/Configurations/10-main.conf
+++ b/Configurations/10-main.conf
@@ -1774,6 +1774,53 @@ my %targets = (
 	inherit_from     => [ "Cygwin-x86" ]
     },
 
+#### MSYS
+    "Msys-common" => {
+        inherit_from     => [ "BASE_unix" ],
+        template         => 1,
+
+        CC               => "gcc",
+        CFLAGS           => picker(default => "-Wall",
+                                   debug   => "-g -O0",
+                                   release => "-O3"),
+        lib_cppflags     => "-DTERMIOS -DL_ENDIAN",
+        sys_id           => "MSYS",
+        thread_scheme    => "pthread",
+        dso_scheme       => "dlfcn",
+        shared_target    => "msys-shared",
+        shared_cppflags  => "-D_WINDLL",
+
+        perl_platform    => 'Msys',
+    },
+    "Msys-x86" => {
+        inherit_from     => [ "Msys-common" ],
+        CFLAGS           => add(picker(release => "-O3 -fomit-frame-pointer")),
+        bn_ops           => "BN_LLONG",
+        perlasm_scheme   => "coff"
+    },
+    "Msys-x86_64" => {
+        inherit_from     => [ "Msys-common" ],
+        bn_ops           => "SIXTY_FOUR_BIT_LONG",
+        perlasm_scheme   => "mingw64"
+    },
+    # Backward compatibility for those using this target
+    "Msys" => {
+	inherit_from     => [ "Msys-x86" ]
+    },
+    # In case someone constructs the Msys target name themself
+    "Msys-i386" => {
+	inherit_from     => [ "Msys-x86" ]
+    },
+    "Msys-i486" => {
+	inherit_from     => [ "Msys-x86" ]
+    },
+    "Msys-i586" => {
+	inherit_from     => [ "Msys-x86" ]
+    },
+    "Msys-i686" => {
+	inherit_from     => [ "Msys-x86" ]
+    },
+
 ##### MacOS X (a.k.a. Darwin) setup
     "darwin-common" => {
         inherit_from     => [ "BASE_unix" ],
diff --git a/Configurations/platform/Msys.pm b/Configurations/platform/Msys.pm
new file mode 100644
index 0000000..33c845d
--- /dev/null
+++ b/Configurations/platform/Msys.pm
@@ -0,0 +1,23 @@
+package platform::Msys;
+
+use strict;
+use warnings;
+use Carp;
+
+use vars qw(@ISA);
+
+require platform::mingw;
+@ISA = qw(platform::mingw);
+
+# Assume someone set @INC right before loading this module
+use configdata;
+
+sub sharedname {
+    my $class = shift;
+    my $lib = platform::mingw->sharedname(@_);
+    $lib =~ s|^lib|msys-| if defined $lib;
+    return $lib;
+}
+
+1;
+
diff --git a/Configurations/shared-info.pl b/Configurations/shared-info.pl
index caf6f90..7fc7da2 100644
--- a/Configurations/shared-info.pl
+++ b/Configurations/shared-info.pl
@@ -55,6 +55,10 @@ my %shared_info;
         shared_ldflag         => '-shared -Wl,--enable-auto-image-base',
         shared_impflag        => '-Wl,--out-implib=',
     },
+    'msys-shared' => {
+        shared_ldflag         => '-shared -Wl,--enable-auto-image-base',
+        shared_impflag        => '-Wl,--out-implib=',
+    },
     'mingw-shared' => sub {
         return {
             %{$shared_info{'cygwin-shared'}},
diff --git a/Configurations/unix-Makefile.tmpl b/Configurations/unix-Makefile.tmpl
index dcd41a9..cd417de 100644
--- a/Configurations/unix-Makefile.tmpl
+++ b/Configurations/unix-Makefile.tmpl
@@ -8,7 +8,7 @@
      our $makedep_scheme = $config{makedep_scheme};
      our $makedepcmd = platform->makedepcmd();
 
-     sub windowsdll { $config{target} =~ /^(?:Cygwin|mingw)/ }
+     sub windowsdll { $config{target} =~ /^(?:Cygwin|Msys|mingw)/ }
 
      # Shared AIX support is special. We put libcrypto[64].so.ver into
      # libcrypto.a and use libcrypto_a.a as static one.
diff --git a/apps/build.info b/apps/build.info
index 020d129..0069a55 100644
--- a/apps/build.info
+++ b/apps/build.info
@@ -76,7 +76,7 @@ IF[{- !$disabled{apps} -}]
   # always depend on a changed configuration.
   DEPEND[progs.c]=../configdata.pm
 
-  IF[{- $config{target} =~ /^(?:Cygwin|mingw|VC-|BC-)/ -}]
+  IF[{- $config{target} =~ /^(?:Cygwin|Msys|mingw|VC-|BC-)/ -}]
     GENERATE[openssl.rc]=../util/mkrc.pl openssl
     SOURCE[openssl]=openssl.rc
   ENDIF
diff --git a/build.info b/build.info
index d9e3c90..1b1293c 100644
--- a/build.info
+++ b/build.info
@@ -91,7 +91,7 @@ IF[{- defined $target{shared_defflag} -}]
   DEPEND[libcrypto.ld libssl.ld]=configdata.pm util/perl/OpenSSL/Ordinals.pm
 ENDIF
 
-IF[{- $config{target} =~ /^(?:Cygwin|mingw|VC-|BC-)/ -}]
+IF[{- $config{target} =~ /^(?:Cygwin|Msys|mingw|VC-|BC-)/ -}]
   GENERATE[libcrypto.rc]=util/mkrc.pl libcrypto
   GENERATE[libssl.rc]=util/mkrc.pl libssl
   DEPEND[libcrypto.rc libssl.rc]=configdata.pm
diff --git a/crypto/build.info b/crypto/build.info
index 2d5b22f..11f08ab 100644
--- a/crypto/build.info
+++ b/crypto/build.info
@@ -140,6 +140,6 @@ GENERATE[riscv64cpuid.s]=riscv64cpuid.pl
 GENERATE[riscv32cpuid.s]=riscv32cpuid.pl
 
 GENERATE[loongarch64cpuid.s]=loongarch64cpuid.pl
-IF[{- $config{target} =~ /^(?:Cygwin|mingw|VC-|BC-)/ -}]
+IF[{- $config{target} =~ /^(?:Cygwin|Msys|mingw|VC-|BC-)/ -}]
   SHARED_SOURCE[../libcrypto]=dllmain.c
 ENDIF
diff --git a/providers/build.info b/providers/build.info
index 3b211bb..e420b0f 100644
--- a/providers/build.info
+++ b/providers/build.info
@@ -120,7 +120,7 @@ IF[{- !$disabled{fips} -}]
   DEPEND[fipsmodule.cnf]=$FIPSGOAL
 
   # Add VERSIONINFO resource for windows
-  IF[{- $config{target} =~ /^(?:Cygwin|mingw|VC-|BC-)/ -}]
+  IF[{- $config{target} =~ /^(?:Cygwin|Msys|mingw|VC-|BC-)/ -}]
     GENERATE[fips.rc]=../util/mkrc.pl fips
     SOURCE[$FIPSGOAL]=fips.rc
   ENDIF
@@ -154,7 +154,7 @@ IF[{- !$disabled{legacy} -}]
     ENDIF
 
     # Add VERSIONINFO resource for windows
-    IF[{- $config{target} =~ /^(?:Cygwin|mingw|VC-|BC-)/ -}]
+    IF[{- $config{target} =~ /^(?:Cygwin|Msys|mingw|VC-|BC-)/ -}]
       GENERATE[legacy.rc]=../util/mkrc.pl legacy
       SOURCE[$LEGACYGOAL]=legacy.rc
     ENDIF
diff --git a/tools/c_rehash.in b/tools/c_rehash.in
index 4dd1b44..8fe3d80 100644
--- a/tools/c_rehash.in
+++ b/tools/c_rehash.in
@@ -21,6 +21,7 @@ my $x509hash = "-subject_hash";
 my $crlhash = "-hash";
 my $verbose = 0;
 my $symlink_exists=eval {symlink("",""); 1};
+if ($^O eq "msys") { $symlink_exists=0 };
 my $removelinks = 1;
 
 ##  Parse flags.
diff --git a/util/perl/OpenSSL/config.pm b/util/perl/OpenSSL/config.pm
index 047902c..b6f56d2 100755
--- a/util/perl/OpenSSL/config.pm
+++ b/util/perl/OpenSSL/config.pm
@@ -164,6 +164,7 @@ my $guess_patterns = [
     [ 'MINGW64.*?:.*?:.*?:x86_64',  '${MACHINE}-whatever-mingw64' ],
     [ 'MINGW.*',                    '${MACHINE}-whatever-mingw' ],
     [ 'CYGWIN.*',                   '${MACHINE}-pc-cygwin' ],
+    [ 'MSYS.*',                     '${MACHINE}-pc-msys' ],
     [ 'vxworks.*',                  '${MACHINE}-whatever-vxworks' ],
 
     # The MACHINE part of the array POSIX::uname() returns on VMS isn't
diff --git a/crypto/bio/bss_dgram.c b/crypto/bio/bss_dgram.c
index 69f3b48..7df025f 100644
--- a/crypto/bio/bss_dgram.c
+++ b/crypto/bio/bss_dgram.c
@@ -1192,7 +1192,7 @@ static int pack_local(BIO *b, MSGHDR_TYPE *mh, const BIO_ADDR *local) {
         cmsg->cmsg_type  = IP_PKTINFO;
 
         info = (struct in_pktinfo *)BIO_CMSG_DATA(cmsg);
-#   if !defined(OPENSSL_SYS_WINDOWS) && !defined(OPENSSL_SYS_CYGWIN) && !defined(__FreeBSD__)
+#   if !defined(OPENSSL_SYS_WINDOWS) && !defined(OPENSSL_SYS_CYGWIN) && !defined(__FreeBSD__) && !defined(OPENSSL_SYS_MSYS)
         info->ipi_spec_dst      = local->s_in.sin_addr;
 #   endif
         info->ipi_addr.s_addr   = 0;
diff --git a/include/openssl/e_os2.h b/include/openssl/e_os2.h
index f1e1795..412cfff 100644
--- a/include/openssl/e_os2.h
+++ b/include/openssl/e_os2.h
@@ -54,7 +54,7 @@ extern "C" {
 #  undef OPENSSL_SYS_UNIX
 #  define OPENSSL_SYS_WIN32_UWIN
 # else
-#  if defined(__CYGWIN__) || defined(OPENSSL_SYS_CYGWIN)
+#  if defined(__CYGWIN__) || defined(OPENSSL_SYS_CYGWIN) || defined(OPENSSL_SYS_MSYS)
 #   define OPENSSL_SYS_WIN32_CYGWIN
 #  else
 #   if defined(_WIN32) || defined(OPENSSL_SYS_WIN32)