From 716d8ad8006bac43e43729c7a504a1ee084e3dc1 Mon Sep 17 00:00:00 2001
From: Alexpux <alexey.pawlow@gmail.com>
Date: Sat, 2 Nov 2013 12:42:49 +0400
Subject: [PATCH] MSYS2 support for cURL

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>
---
 m4/curl-compilers.m4      |    2 +-
 m4/xc-lt-iface.m4         |    2 +-
 2 files changed, 1644 insertions(+), 2 deletions(-)

diff --git a/m4/curl-compilers.m4 b/m4/curl-compilers.m4
index dc6d3aa..fbe235d 100644
--- a/m4/curl-compilers.m4
+++ b/m4/curl-compilers.m4
@@ -861,7 +861,7 @@ AC_DEFUN([CURL_SET_COMPILER_WARNING_OPTS], [
             CURL_ADD_COMPILER_WARNINGS([tmp_CFLAGS], [enum-conversion])
             CURL_ADD_COMPILER_WARNINGS([tmp_CFLAGS], [sometimes-uninitialized])
             case $host_os in
-              cygwin* | mingw*)
+              cygwin* | msys* | mingw*)
                 dnl skip missing-variable-declarations warnings for Cygwin and
                 dnl MinGW because the libtool wrapper executable causes them
                 ;;
diff --git a/m4/xc-lt-iface.m4 b/m4/xc-lt-iface.m4
index a408800..2933955 100644
--- a/m4/xc-lt-iface.m4
+++ b/m4/xc-lt-iface.m4
@@ -265,7 +265,7 @@ elif test "x$allow_undefined_flag" = 'xunsupported'; then
   xc_lt_shlib_use_no_undefined='yes'
 fi
 case $host_os in @%:@ (
-  cygwin* | mingw* | cegcc* | os2* | aix*)
+  cygwin* | msys* | mingw* | cegcc* | os2* | aix*)
     xc_lt_shlib_use_no_undefined='yes'
     ;;
 esac
